client/protocol.go

```diff
@@ -0,0 +1,413 @@
+			}
+			proposeRes = <-slowResCh
+			err = <-slowErrCh
+			return proposeRes, err
+		case proposeRes := <-slowResCh:
+			err := <-slowErrCh
+			return proposeRes, err
+		}
+	} else {
+		<-fastResCh
+		<-fastErrCh
+		proposeRes := <-slowResCh
+		err := <-slowErrCh
+		return proposeRes, err
+	}
+}
+
+// The fast round of Curp protocol
+// It broadcast the requests to all the curp servers.
+func (c protocolClient) fastRound(cmd *xlineapi.Command) (*ProposeResponse, error) {
+	logrus.Info("Fast round start. Propose ID:", cmd.ProposeId)
+
+	isReceivedLeaderReq := false
+	isReceivedSuccessRes := 0
+	okCnt := 0
+	var clients []curpapi.ProtocolClient
+	var cmdRes xlineapi.CommandResponse
+	var proposeErr xlineapi.ExecuteError
+	proposeCh := make(chan *curpapi.ProposeResponse)
+
+	for _, m := range c.addrs {
+		conn, err := grpc.Dial(m, grpc.WithTransportCredentials(insecure.NewCredentials()), grpc.WithIdleTimeout(c.timeout.idleTimeout))
+		if err != nil {
+			logrus.Warn(err)
+		} else {
+			protocolClient := curpapi.NewProtocolClient(conn)
+			clients = append(clients[:], protocolClient)
+		}
+	}
+
+	serializeCmd, err := proto.Marshal(cmd)
+	if err != nil {
+		return nil, errors.New("command marshal fail")
+	}
+
+	for _, protocolClient := range clients {
+		client := protocolClient
+		go func() {
+			ctx, cancel := context.WithTimeout(context.Background(), c.timeout.proposeTimeout)
+			defer cancel()
+			res, err := client.Propose(ctx, &curpapi.ProposeRequest{Command: serializeCmd})
+			if err != nil {
+				logrus.Debug(err)
+				proposeCh <- nil
+			} else {
+				proposeCh <- res
+			}
+		}()
+	}
+
+	for i := 0; i < len(clients); i++ {
+		proposeRes := <-proposeCh
+
+		if proposeRes == nil {
+			continue
+		}
+
+		if proposeRes.ExeResult != nil {
+			isReceivedLeaderReq = true
+		}
+
+		switch proposeResp := proposeRes.ExeResult.(type) {
+		case *curpapi.ProposeResponse_Result:
+			switch cmdResult := proposeResp.Result.Result.(type) {
+			case *curpapi.CmdResult_Er:
+				err := proto.Unmarshal(cmdResult.Er, &cmdRes)
+				if err != nil {
+					logrus.Debug(err)
+				} else {
+					isReceivedSuccessRes = 1
+					okCnt++
+				}
+			case *curpapi.CmdResult_Error:
+				err := proto.Unmarshal(cmdResult.Error, &proposeErr)
+				if err != nil {
+					logrus.Debug(err)
+				} else {
+					isReceivedSuccessRes = 2
+					okCnt++
+				}
+			default:
+				logrus.Warn("unknown command result type")
+			}
+		case *curpapi.ProposeResponse_Error:
+			logrus.Debug(proposeResp.Error)
+		default:
+			okCnt++
+		}
+
+		if isReceivedLeaderReq && okCnt >= superQuorum(len(c.connects)) {
+			logrus.Info("Fast round success. ProposeID:", cmd.ProposeId)
+			if isReceivedSuccessRes == 1 {
+				return &ProposeResponse{
+					CommandResp:        &cmdRes,
+					IsFastRoundSuccess: true,
+				}, nil
+			}
+			if isReceivedSuccessRes == 2 {
+				return &ProposeResponse{
+						IsFastRoundSuccess: true,
+					}, ProposeError{
+						ExecuteError: &proposeErr,
+					}
```

Comment:
when `err != nil`, the other part of return value is unnecessary.

Reply:
done